common/domain/handler.go

```diff
@@ -546,102 +542,59 @@ func (d *handlerImpl) handleFailoverRequest(ctx context.Context,
-	isActiveActive := intendedDomainState.ReplicationConfig.IsActiveActive()
-
-	// case 1. active-passive domain's ActiveClusterName is changed
-	if !wasActiveActive && !isActiveActive {
-		intendedDomainState.FailoverVersion = d.clusterMetadata.GetNextFailoverVersion(
-			intendedDomainState.ReplicationConfig.ActiveClusterName,
-			currentState.FailoverVersion,
-			updateRequest.Name,
-		)
+	intendedDomainState.FailoverVersion = d.clusterMetadata.GetNextFailoverVersion(
+		intendedDomainState.ReplicationConfig.ActiveClusterName,
+		currentState.FailoverVersion,
+		updateRequest.Name,
+	)
 
-		d.logger.Debug("active-passive domain failover",
-			tag.WorkflowDomainName(intendedDomainState.Info.Name),
-			tag.Dynamic("failover-version", intendedDomainState.FailoverVersion),
-			tag.Dynamic("failover-type", failoverType),
-		)
+	// this is an intended failover step, to capture the current cluster's domain configuration
+	// (represented by the notification-version counter) and set it when running failover to allow
+	// systems like graceful failover to dedup failover processes and callbacks
+	intendedDomainState.FailoverNotificationVersion = notificationVersion
 
-		err = updateFailoverHistoryInDomainData(intendedDomainState.Info, d.config, NewFailoverEvent(
-			now,
-			failoverType,
-			&currentActiveCluster,
-			updateRequest.ActiveClusterName,
-			nil,
-			nil,
-		))
-		if err != nil {
-			d.logger.Warn("failed to update failover history", tag.Error(err))
-		}
-	}
+	isActiveActive := intendedDomainState.ReplicationConfig.IsActiveActive()
 
-	// case 2. active-passive domain is being migrated to active-active
-	if !wasActiveActive && isActiveActive {
-		// for active-passive to active-active migration,
-		// we increment failover version so top level failoverVersion is updated and domain data is replicated.
+	if wasActiveActive || isActiveActive {
+		// if the domain was ever active-active, we bump the failover-version at the domain level
+		// to ensure that it gets incremented every single time, even though the failover may be only
+		// at the cluster-attribute level (ie the FailoverVersion may just go up by the failover-increment
+		// but otherwise be a noop - the actual active cluster may not have changed)).
+		//
+		// The reason for this is caution and simplicity at the time of writing:
+		// It's harmless to bump and this simplifies thinking about failover events
+		// through out the rest of the Cadence codebase. We can reliably assume that every failover
+		// event that has meaningful changes only needs to subscribe to this fencing token.
+		// to detect changes.
+		//
+		// Any parts of the system (domain-callbacks, domain cache etc) that watch for
+		// failover events and may trigger some action or cache invalidation will be watching
+		// the domain-level failver counter for changes. Therefore, by bumping it even for
+		// cases where it isn't changing, we ensure all these other subprocesses will
```

Comment:
I still think we should find the subprocesses depending on this and extend them to support cluster-attribute-level failover instead of increasing the domain level failover version fall all failovers. 
But I'm ok with the current state.